r1cs/gadgets/std/src/bits/macros.rs

```diff
@@ -0,0 +1,2649 @@
+                        allocated_bits.extend_from_slice(fe_bits.as_slice());
+                    }
+
+                    // pad with additional zero bits to have a number of bits which is multiple of $bit_size
+                    while allocated_bits.len() % $bit_size != 0 {
+                        allocated_bits.push(Boolean::constant(false));
+                    }
+
+                    // Chunk up slices of $bit_size bits into bytes.
+                    Ok(allocated_bits[..]
+                        .chunks($bit_size)
+                        .enumerate()
+                        .map(|(i, chunk)| Self::from_bits_le(cs.ns(|| format!("pack input chunk {}", i)), chunk))
+                        .collect::<Result<_, SynthesisError>>()?)
+                }
+
+                /// Allocates a vector of `Self` from a slice of values of $native_type by serializing
+                /// them to sequence of bytes and then converting them to a sequence of `ConstraintF`
+                /// elements, (thus reducing the number of input allocations);
+                ///Then,  this list of `ConstraintF` gadgets is converted back into
+                /// bits and then packs chunks of such into `Self`.
+                pub fn alloc_input_vec<ConstraintF, CS, T>(
+                    cs: CS,
+                    values: &[T],
+                ) -> Result<Vec<Self>, SynthesisError>
+                where
+                    ConstraintF: PrimeField,
+                    CS: ConstraintSystemAbstract<ConstraintF>,
+                    T: Into<$native_type> + Copy,
+                {
+                    // convert values to vector of bytes
+                    let mut values_to_bytes = Vec::new();
+                    for val in values {
+                        let val_bytes = Into::<$native_type>::into(*val).to_le_bytes();
+                        values_to_bytes.extend_from_slice(&val_bytes[..]);
+                    }
+                    // alloc vector of `Self` from vector of bytes, minimizing the number of
+                    // `ConstraintF` elements allocated
+                    Self::alloc_input_vec_from_bytes(cs, values_to_bytes.as_slice())
+                }
+
+                /// Construct a constant vector of `Self` from a vector of `$native_type`
+                pub fn constant_vec<T: Into<$native_type> + Copy>(values: &[T]) -> Vec<Self> {
+                    let mut result = Vec::new();
+                    for val in values {
+                        result.push(Self::constant((*val).into()));
+                    }
+                    result
+                }
+
+                /// enforces that self >= other. This function is provided as it is much efficient
+                /// in terms of constraints with respect to the default implementation of
+                /// ComparisonGadget, which relies on the smaller_than functions
+                pub fn enforce_greater_or_equal_than<ConstraintF, CS>(&self, mut cs: CS, other: &Self)
```

Comment:
Why do you think that `sub_noborrow` does not affect the constraint system? The last operation of the function is calling enforce equal. 
Note that the function does not take directly the constraint system `cs` as input because for efficiency it employs the `MultiEq` constrain system, which however wraps `cs` as the `Self::Root` constraint system, that is the constraint system over which constraints built by `MultiEq` are enforced.

Reply:
Ok